### DISCUSSION TITLE
changing the color of the title bar?

### DISCUSSION THREAD

**Discussion by clear-code-projects** [TomSchimansky/CustomTkinter#1011] [General]
Hi, I wonder if it is possible to change the title of the titlebar? I couldn't find an answer in the documentation but since it is different from the default maybe there is a way to do it easily? 

(would be a really good feature, by the way)

**Comment by harshnagar** [comment]
You are Talking about this Title Bar of the tkinter windows ?
<img width="321" alt="tital" src="https://user-images.githubusercontent.com/35253819/211160014-505a17da-756f-46ca-876e-65e540610bad.png">

**Comment by sumeshir26** [comment]
@clear-code-projects I think that is possible using the DWM Api from the `ctypes.windll.dwmapi`, similar to how the Windows 11 mica effect is set on legacy tkinter/QT applications over here: https://github.com/marticliment/win32mica/blob/ba5ca37161c78712a98bf5ebea3a90882d11c903/src/win32mica/__init__.py#L11

Btw I reallay enjoy your videos :)

**Comment by gwiggen1** [comment]
If you are talking about the window title:
```
root.title("")
```
or 
```
main.title("")
```
it is built in

**Comment by Akascape** [comment]
@clear-code-projects 
After searching the whole web on this topic, I finally found the solution for this:
We have to use  `ctypes` in order to do this. There is a thing called [DwmSetWindowAttribute](https://learn.microsoft.com/en-us/windows/win32/api/dwmapi/ne-dwmapi-dwmwindowattribute) under `windll.dwmapi` which will allow us to do changes with window properties which is managed by the OS.

### Here is the final result:
![CODE](https://user-images.githubusercontent.com/89206401/211208553-77828ae7-3daf-4e98-a9ff-9b32350d269e.jpg)

## CODE
```python
import customtkinter
from ctypes import windll, byref, sizeof, c_int

root = customtkinter.CTk()
root.update()

HWND = windll.user32.GetParent(root.winfo_id()) # the window we want to change

"""
DWMWA_ATTRIBUTES (for windows 11 title bar) 
CAPTION COLOR (HEADER) = 35
BORDER COLOR = 34
TITLE COLOR = 36
"""

DWMWA_ATTRIBUTE = 35

COLOR = 0x00FF0000 # color should be in hex order: 0x00bbggrr

windll.dwmapi.DwmSetWindowAttribute(HWND, DWMWA_ATTRIBUTE, byref(c_int(COLOR)), sizeof(c_int))

root.mainloop()
```
Make sure that the color can be customized based on the hex values in this order: `0x00bbggrr`. You can find more about this color api here: https://learn.microsoft.com/en-us/windows/win32/gdi/colorref

**You can also customize other parameters like border color, title color and get a final look like this:**
![Screenshot 2023-01-08 222953](https://user-images.githubusercontent.com/89206401/211209161-30d1e84b-2f21-49da-8788-0f16923c60fa.jpg)

Other themes can also be applied like the **mica** one:
`windll.dwmapi.DwmSetWindowAttribute(HWND, 1029, byref(c_int(0x01)), sizeof(c_int))`

That's all I found. Thanks!

**Reply by clear-code-projects** [reply]
What a hero, thank you so much!

**Reply by clear-code-projects** [reply]
Quick follow up: I guess you wouldn't have an idea how to do this on macOS to keep things consistent? I've been looking around for half an hour now and just cannot find anything remotely useful :(

**Reply by Akascape** [reply]
@clear-code-projects I am not a mac user, so I don't know much about its inner system and mechanisms. 
Ig there must be a way to change the header with mac also. (maybe with some different APIs). 

But atleast windows users can enjoy now.

**Reply by Akascape** [reply]
Here is the method ctk uses to change the title bar:
https://github.com/TomSchimansky/CustomTkinter/blob/master/customtkinter/windows/ctk_tk.py#L223-L228

**Reply by clear-code-projects** [reply]
It was a long shot but thank you so much, you have been super helpful :)

**Reply by Xiaokang2022** [reply]
I've spent hours searching the entire internet for answers, and only yours is the answer I expected! It's amazing! Thank you!

**Reply by Akascape** [reply]
> I've spent hours searching the entire internet for answers, and only yours is the answer I expected! It's amazing! Thank you!

@Xiaokang2022 I recommend you to use my pywinstyles package to make the work easier. You just need to add one line of code in order to change the title bar color 

https://github.com/Akascape/py-window-styles
